x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/registry/ModelRegistry.java

```diff
@@ -306,7 +314,139 @@ private ModelConfigMap createModelConfigMap(SearchHits hits, String inferenceEnt
         );
     }
 
+    public void updateModelTransaction(Model newModel, Model existingModel, ActionListener<Boolean> finalListener) {
+
+        String inferenceEntityId = newModel.getConfigurations().getInferenceEntityId();
+        logger.info("Attempting to store update to inference endpoint [{}]", inferenceEntityId);
+
+        if (preventDeletionLock.contains(inferenceEntityId)) {
+            logger.warn(format("Attempted to update endpoint [{}] that is already being updated", inferenceEntityId));
+            finalListener.onFailure(
+                new ElasticsearchStatusException(
+                    "Endpoint [{}] is currently being updated. Try again once the update completes",
+                    RestStatus.CONFLICT,
+                    inferenceEntityId
+                )
+            );
+            return;
+        } else {
+            preventDeletionLock.add(inferenceEntityId);
+        }
+
+        SubscribableListener.<BulkResponse>newForked((subListener) -> {
+            // in this block, we try to update the stored model configurations
+            IndexRequest configRequest = createIndexRequest(
+                Model.documentId(inferenceEntityId),
+                InferenceIndex.INDEX_NAME,
+                newModel.getConfigurations(),
+                true
+            );
+
+            ActionListener<BulkResponse> storeConfigListener = subListener.delegateResponse((l, e) -> {
+                // this block will only be called if the bulk unexpectedly throws an exception
+                preventDeletionLock.remove(inferenceEntityId);
+                l.onFailure(e);
+            });
+
+            client.prepareBulk().add(configRequest).setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE).execute(storeConfigListener);
+
+        }).<BulkResponse>andThen((subListener, configResponse) -> {
+            // in this block, we respond to the success or failure of updating the model configurations, then try to store the new secrets
+            if (configResponse.hasFailures()) {
+                // if storing the model configurations failed, it won't throw an exception, we need to check the BulkResponse and handle the
+                // exceptions ourselves.
+                logger.error(
+                    format("Failed to update inference endpoint [%s] due to [%s]", inferenceEntityId, configResponse.buildFailureMessage())
+                );
+                // Since none of our updates succeeded at this point, we can simply return.
+                finalListener.onFailure(
```

Comment:
Do we also need to unlock here as well?

Reply:
Yea, I think that's correct. We probably should be unlocking whenever we call the finalListener or the the final sub listener. Do you think you could add this tomorrow? If not, I can probably do it at some point before ff.